sfdx/index.js

```diff
@@ -0,0 +1,111 @@
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+var SDK = require('./shared/constants'),
+    createHelper = require('./shared/createHelper'),
+    configHelper = require('./shared/configHelper'),
+    logError = require('./shared/utils').logError;
+
+// Flags
+function getFlags(cli, commandName) {
+    return configHelper.getArgsExpanded(cli, commandName);
+}
+
+// Validation
+function validateCommand(cli, commandName, vals) {
+    var success = true;
+    var args = configHelper.getArgsExpanded(cli, commandName);
+    for (var arg of args) {
+        var val = vals[arg.name];
+
+        if (typeof arg.validate === 'function' && !arg.validate(val, cli)) {
+            success = false;
+            logError(arg.error(val, cli));
+        }
+    }
+    return success;
+}
+
+// Run command
+function runCommand(cli, commandName, vals) {
+    switch(commandName) {
+    case SDK.commands.create.name:
+    case SDK.commands.createWithTemplate.name:
+        createHelper.createApp(cli, vals);
+        break;
+    case SDK.commands.version.name:
+        configHelper.printVersion(cli);
+        break;
+    }
+}
+
+// Topics
+function getTopics() {
+    var topics = [];
+    for (var cliName in SDK.forceclis) {
+        var cli = SDK.forceclis[cliName];
+        topics.push({
+            name: cli.sfx_topic,
+            description:cli.sfdx_description
+        });
+    }
+    return topics;
+}
+
+// Commands
+function getCommands() {
+    var commands = [];
+    for (var cliName in SDK.forceclis) {
+        var cli = SDK.forceclis[cliName];
+        for (var commandName of cli.commands) {
+            var command = configHelper.getCommandExpanded(cli, commandName);
+            commands.push({
+                cli: cli,
+                topic: cli.sfdx_topic,
+                command: commandName,
+                description: command.description,
+                flags: command.args,
+                run(context) {
+                    if (validateCommand(this.cli, this.command, context.flags)) {
+                        runCommand(this.cli, this.command, context.flags);
+                    }
+                }
+            });
+        }
+    }
+    return commands;
+}
+
+module.exports = {
```

Comment:
An sfdx (or heroku) plugin needs to export a namespace, topics and commands.
We build them all here, relying heavily on constants.js.